ecommerce/extensions/payment/views/liqpay.py

```diff
@@ -0,0 +1,169 @@
+                            basket=basket).exclude(transaction_id__isnull=True).exclude(transaction_id='').exists():
+                        logger.info(
+                            'Received LiqPay payment notification for basket [%d] which is associated '
+                            'with existing order [%s] or had an existing valid payment notification. '
+                            'No payment was collected, and no new order will be created.',
+                            basket.id,
+                            order_number
+                        )
+                    else:
+                        logger.info(
+                            'Received duplicate LiqPay payment notification for basket [%d] which is not associated '
+                            'with any existing order (Missing Order Issue)',
+                            basket.id,
+                        )
+                    return redirect(self.payment_processor.error_url)
+                except PaymentError:
+                    logger.exception(
+                        'LiqPay payment failed for basket [%d]. The payment response was recorded in entry [%d].',
+                        basket.id,
+                        ppr.id
+                    )
+                    return redirect(self.payment_processor.error_url)
+        except:  # pylint: disable=bare-except
+            logger.exception('Attempts to handle payment for basket [%d] failed.', basket.id)
+            return redirect(receipt_url)
+
+        try:
+            # Note (CCB): In the future, if we do end up shipping physical products, we will need to
+            # properly implement shipping methods. For more, see
+            # http://django-oscar.readthedocs.org/en/latest/howto/how_to_configure_shipping.html.
+            shipping_method = NoShippingRequired()
+            shipping_charge = shipping_method.calculate(basket)
+
+            # Note (CCB): This calculation assumes the payment processor has not sent a partial authorization,
+            # thus we use the amounts stored in the database rather than those received from the payment processor.
+            user = basket.owner
+            order_total = OrderTotalCalculator().calculate(basket, shipping_charge)
+
+            order = self.handle_order_placement(
+                order_number=order_number,
+                user=user,
+                basket=basket,
+                shipping_address=None,
+                shipping_method=shipping_method,
+                shipping_charge=shipping_charge,
+                billing_address=None,
+                order_total=order_total,
+                request=request
+            )
+
+            self.handle_post_order(order)
+            return redirect(receipt_url)
+        except Exception as e:  # pylint: disable=broad-except
+            logger.exception(self.order_placement_failure_msg, basket.id, e)
+            return redirect(receipt_url)
```

Comment:
I wish to move `return redirect(receipt_url)` to `finally` section at least.

Reply:
refactored